teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +161,574 @@ alpha features. This is not a requirement for Tekton, because even in case of
+
+Cons:
+* There would be a lot of fields (once we promote a field, we'd need to maintain the previous versions
+  for some amount of time)
+
+### Add alpha section(s) to the spec
+
+Similar to the previous option but instead of using field names we create a section in each CRD.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: task-with-parameters
+spec:
+  steps:
+    - name: this-can-fail
+      image: some-image
+      args: ["some-arg"]
+      alpha: # the alpha stability section for a step
+          allow_failure: true # new field
+```
+
+Or we could try to make one alpha section only per CRD:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: task-with-parameters
+spec:
+  steps:
+    - name: this-can-fail
+      image: some-image
+      args: ["some-arg"]
+  alpha: # the alpha stability section - might get a bit complicated
+    steps: # i guess this augments the step section above?
+    - name: this-can-fail
+      allow_failure: true # new field
+```
+
+OR we could duplicate everything in an alpha section:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: task-with-parameters
+spec:
+  alpha: # this section is the whole spec duplicated PLUS alpha fields
+    steps:
+    - name: this-can-fail
+      image: some-image
+      args: ["some-arg"]
+      allow_failure: true # new field
+```
+
+#### Pros and cons
+
+Pros:
+* Very clear when a field is alpha and that changes will be required once it is stable
+
+Cons:
+* Similar to previous option, a lot of of duplication
+* Confusing to reason about what happens if you combine the alpha section with the non-alpha sections
+
+### Build separate release with alpha features
+
+Build multiple releases: one which only has beta fields in beta types, etc.,
+and one that contains alpha fields as well.
+
+#### Pros and cons
+
+Pros:
+* Clear when you're opting in and when you're not
+
+Cons:
+* Opting in is harder (install something entirely different)
+* Additional complication in our release process (esp if we need to make several different versions)
+
+## Implementation
+
+In order to consider this TEP implemented we will need to:
+
+* Create clear documentation showing contributors how to add new features i.e. how do you add a field that is
+  guarded by these flags including:
+  * Verifying in the admission controller that the flag is set if the field is used
+  * How to add the appropriate [tests](#test-plan)
+  * How to correctly [document](#docs)) the feature
+* Add [documentation](#docs) about the stability level of each field
+* Add documentation about how to use the new flag
+* Add [tests](#test-plan) i.e. initially an end to end (or example) test that sets the flag to `alpha`
+* Deprecate the `enable-custom-tasks` flag (can be implied by `enable-api-fields:alpha`) - possibly `enable-oci-bundles`
+  as well
+
```

Comment:
@sbwsg I wanted to give you a heads up for what might be involved for your steps + sidecar workspaces PR - but i dont think YOU need to take all of this on, im thinking you'd need to make sure your feature is in an "alpha" docs section, and we'd need to make sure there's a test or example that enables the flag